File: backend/jobs.py
"""Background jobs a user started, with progress a browser can watch and a cancel that works.

WHY THIS EXISTS
    Every long operation on the Management Dashboard already ran as `threading.Thread` + `Queue`,
    with the SSE endpoint draining the queue directly. That shape has one defect and it is not
    cosmetic: **the client is not attached to the work**. Disconnect — navigate away, close the
    tab, press a Cancel that aborts the fetch — and the generator stops being consumed while the
    worker thread runs happily to completion. On the S&P's fundamentals fill that is 237
    constituents x ~3 GuruFocus calls ~= 711 calls that could not be stopped by anything short of
    a redeploy.

    Putting a JOB between the two fixes that, and three other things fall out of it for free:
      * the run survives navigation and reload — the toast re-attaches instead of the work
        becoming invisible-but-still-running;
      * two tabs can watch the same run;
      * several runs can be in flight at once, each reporting separately.

⚠⚠ CANCELLATION IS COOPERATIVE, AND THAT IS WHAT MAKES IT *SAFE*.
    Nothing here kills a thread. `ctx.check()` raises `JobCancelled` at a point the WORKER chose,
    which is always a boundary where the database is already consistent — between two GuruFocus
    feeds, between two companies. A thread killed mid-write would leave exactly the half-loaded
    state this app spends so much effort refusing to display. The cost is latency: pressing Cancel
    during a feed stops the job after that feed, not instantly. That is the right trade and the UI
    says "cancelling…" rather than pretending otherwise.

⚠ A DISCONNECT IS NOT A CANCEL. Closing the stream leaves the job running, deliberately: the
    reader may just be navigating away, and silently abandoning a half-finished ingest because
    someone clicked a different page is how you get a company with statements and no estimates.
    Cancel is an explicit POST and nothing else.

⚠ IN-PROCESS, NOT A TABLE, and that is a considered limit rather than an oversight. It matches the
    assumption `ingest/phases/pipeline.py::_PIPELINE_LOCK` already makes — one instance, with
    `DISABLE_SCHEDULER=1` on any replica — so there is no migration and no second source of truth.
    A deploy mid-job loses the job, but the work dies with the process anyway, so nothing is
    orphaned; what is lost is the record, and `ingest_run` remains where durable history lives for
    the scheduled pipeline. If this ever needs to survive a restart or span replicas, that is the
    point at which it earns a table — not before.
"""
from __future__ import annotations

import logging
import threading
import time
import uuid
from dataclasses import dataclass, field
from typing import Any, Callable

_log = logging.getLogger(__name__)

# How long a finished job stays readable before it is pruned. Long enough that a reader who
# stepped away still sees how it ended; short enough that the registry cannot grow without bound.
RETAIN_SECONDS = 15 * 60

#: How long a RUNNING job may go without emitting anything before it is written off.
#:
#: ⚠⚠ WITHOUT THIS A HUNG WORKER POISONS THE APP UNTIL THE NEXT DEPLOY, and it did (2026-08-31,
#: production). `Refresh all portfolios` blocked inside a Playwright call, so it never reached a
#: `ctx.check()` — Cancel could only ever SET the flag, leaving the card reading "cancelling…" with
#: "starting…" as its last line. Nothing could clear it:
#:
#:   * `_prune` drops TERMINAL jobs only, so the corpse stayed in the registry;
#:   * `attachRunningJobs` adopts every `running` job on page load, so the toast came back on
#:     EVERY visit to /management-dashboard, and only a terminal job ever gets the linger-and-
#:     dismiss countdown — so it could not be got rid of;
#:   * `find_running` matched it, so pressing the button again ATTACHED TO THE CORPSE instead of
#:     starting a run. That is why the card never got past "starting…": there was no new run.
#:
#: ⚠ THIRTY MINUTES, AND THE UNIT IS SILENCE, NOT AGE. A long job is fine — the fleet scan takes
#: many minutes — but every one of them narrates: the AIRS refresh emits per account, a fill emits
#: per company. Half an hour with nothing at all is a worker that is not coming back.
#:
#: ⚠⚠ IT CANNOT KILL THE THREAD, AND THE SUMMARY SAYS SO. Cancellation here is cooperative by
#: design (see the module note); a thread blocked in a vendor call cannot be interrupted from
#: outside. What this reclaims is the REGISTRY — the card can finish, and the next press starts a
#: real run instead of adopting a dead one. The thread, if it ever wakes, finds `_cancel` set.
STALE_SECONDS = 30 * 60

TERMINAL = ("done", "failed", "cancelled")


class JobCancelled(Exception):
    """Raised by `JobCtx.check()` when cancellation was requested. The runner catches it and marks
    the job `cancelled` — a worker should let it propagate rather than swallowing it.

    ⚠ ITS MESSAGE, IF IT HAS ONE, BECOMES THE SUMMARY. A worker that stops part-way usually knows
    something the registry cannot — which account it stopped before, what it had already stored,
    what is now stale — and that is the whole reason a reader looks at a cancelled card. Raised
    bare (as `check()` does) it still reads "cancelled", so nothing had to change to keep working.
    """


@dataclass
class Job:
    id: str
    kind: str
    label: str
    status: str = "running"
    created_at: float = field(default_factory=time.time)
    ended_at: float | None = None
    done: int = 0
    total: int = 0
    #: When this job last said anything. ⚠ THE HEARTBEAT, and it starts at creation rather than at
    #: 0 so a job that dies before its first `emit` still ages out instead of living forever.
    last_event_at: float = field(default_factory=time.time)
    summary: str | None = None
    # ⚠ EXTERNAL, METERED CALLS — NOT A REQUEST COUNT. Our own database reads are free and
    # unlimited; a GuruFocus call comes out of a finite monthly quota, and a reader deciding
    # whether to press a button again deserves to know which kind they just spent. Jobs that spend
    # nothing metered leave this at 0 and the UI shows nothing, which is why it is worth
    # distinguishing "spent none" from "does not apply" only in the sense that both render blank.
    api_calls: int = 0
    # ⚠ AN APPEND-ONLY LIST, NOT A QUEUE, AND THE DIFFERENCE IS THE WHOLE FEATURE. A Queue can be
    # drained by exactly one consumer, once — so a reader who reloads has missed everything, and a
    # second tab steals events from the first. A list plus a per-subscriber cursor gives replay and
    # any number of watchers for free.
    events: list[dict] = field(default_factory=list)
    _cancel: threading.Event = field(default_factory=threading.Event)
    _lock: threading.Lock = field(default_factory=threading.Lock)

    @property
    def terminal(self) -> bool:
        return self.status in TERMINAL

    @property
    def cancel_requested(self) -> bool:
        return self._cancel.is_set()

    def since(self, seq: int) -> list[dict]:
        """Every event after `seq`. Cheap, and safe to call from the event loop while the worker
        thread appends — the lock is held only for the slice."""
        with self._lock:
            return [e for e in self.events if e["seq"] > seq]

    def public(self) -> dict:
        return {
            "id": self.id, "kind": self.kind, "label": self.label, "status": self.status,
            "done": self.done, "total": self.total, "summary": self.summary,
            "api_calls": self.api_calls, "cancel_requested": self.cancel_requested,
            "created_at": self.created_at, "ended_at": self.ended_at,
        }


@dataclass
class JobCtx:
    """What a worker is handed. The only surface a job function needs."""

    job: Job

    @property
    def cancelled(self) -> bool:
        return self.job.cancel_requested

    def check(self) -> None:
        """⚠ CALL THIS AT EVERY BOUNDARY WHERE STOPPING IS SAFE, and nowhere else. It is the only
        thing that makes Cancel real: a worker that never checks simply runs to completion and the
        button is a lie."""
        if self.job.cancel_requested:
            raise JobCancelled

    def emit(self, kind: str, message: str, **data: Any) -> None:
        """One progress line. `done`/`total` are lifted onto the job so a re-attaching client gets
        the bar's position without replaying the whole log."""
        j = self.job
        with j._lock:  # noqa: SLF001
            seq = len(j.events) + 1
            if "done" in data:
                j.done = int(data["done"] or 0)
            if "total" in data:
                j.total = int(data["total"] or 0)
            j.events.append({"seq": seq, "kind": kind, "message": message, **data})
            # ⚠ THE HEARTBEAT IS EVERY LINE, not a separate call a worker must remember to make.
            # Every job on this registry already narrates; asking for a second signal would mean
            # the ones that forgot it get reaped while running.
            j.last_event_at = time.time()

    def progress(self, done: int, total: int, message: str, **data: Any) -> None:
        self.emit("progress", message, done=done, total=total, **data)

    def spent(self, calls: int) -> None:
        """Record metered external calls. Additive, so a worker can report per unit of work.

        ⚠ REPORT IT EVEN WHEN THE JOB THEN FAILS. The quota is gone either way, and a failed run
        that says it cost nothing is the one that gets retried until the month's budget is.
        """
        if calls:
            with self.job._lock:  # noqa: SLF001
                self.job.api_calls += int(calls)


_JOBS: dict[str, Job] = {}
_REGISTRY_LOCK = threading.Lock()


def reap_stalled() -> None:
    """Write off a running job that has gone silent for `STALE_SECONDS` — see that constant.

    ⚠ IT IS A STATUS CHANGE, NOT A KILL. The thread may still be blocked in a vendor call; nothing
    here can interrupt it and the summary does not pretend otherwise. What is reclaimed is the
    registry entry, so the card can reach a terminal state and the next press starts a real run.

    ⚠ `failed`, NOT `cancelled`. Nobody asked for this to stop; it stopped answering. Filing it as
    a cancellation would put a worker's crash in the same column as a reader's decision.

    ⚠ AND `_cancel` IS SET ON THE WAY OUT, so a worker that does eventually wake up unwinds at its
    next `ctx.check()` rather than carrying on writing under a job the registry has written off.

    ⚠ PUBLIC BECAUSE THE SSE STREAM CALLS IT TOO. `_prune` runs on start and on list — neither of
    which a tab that is already WATCHING a hung job will do — so without a tick-level call that
    reader keeps their card forever while everyone else's is reaped. It is a timestamp compare over
    a handful of jobs.
    """
    cutoff = time.time() - STALE_SECONDS
    stalled: list[Job] = []
    with _REGISTRY_LOCK:
        for j in _JOBS.values():
            if not j.terminal and j.last_event_at < cutoff:
                stalled.append(j)
    for j in stalled:
        quiet = int((time.time() - j.last_event_at) / 60)
        j._cancel.set()  # noqa: SLF001
        j.status = "failed"
        j.summary = (f"No progress for {quiet} minutes — abandoned. The worker may still be "
                     "running; nothing here can stop a thread that is not asking to be stopped.")
        j.ended_at = time.time()
        _log.warning("[job] %s (%s) abandoned after %s minutes of silence", j.label, j.kind, quiet)
        # ⚠ AN EVENT, SO A WATCHING TAB LEARNS WHY. A card that flips to `failed` with no line
        # explaining it is the same dead end from the reader's side.
        JobCtx(j).emit("error", j.summary)


def _prune() -> None:
    """Reap the stalled, then drop finished jobs past `RETAIN_SECONDS`. Called on every start and
    list, so the registry is tidied by use rather than by a timer nobody would remember exists.

    ⚠ THE ORDER MATTERS: reaping makes a job terminal, and the drop below only ever removes
    terminal ones — so a stalled job leaves in two steps rather than being stuck for another
    `RETAIN_SECONDS`. It stays readable for the retention window like any other outcome, which is
    what puts one explaining card in front of the reader instead of silence.
    """
    reap_stalled()
    cutoff = time.time() - RETAIN_SECONDS
    with _REGISTRY_LOCK:
        for jid in [j.id for j in _JOBS.values()
                    if j.terminal and (j.ended_at or 0) < cutoff]:
            _JOBS.pop(jid, None)


def find_running(kind: str, label: str) -> Job | None:
    """The live job for this exact piece of work, if there is one.

    ⚠⚠ IT REAPS FIRST, AND WITHOUT THAT THE IDEMPOTENCE BELOW BECOMES A TRAP. `start` attaches to
    whatever this returns instead of launching — right for a run in flight, and fatal for a hung
    one: every press adopts the corpse and nothing new ever starts. Measured in production on
    `Refresh all portfolios`, whose card sat on "starting…" through press after press.
    """
    reap_stalled()
    with _REGISTRY_LOCK:
        for j in _JOBS.values():
            if j.kind == kind and j.label == label and not j.terminal:
                return j
    return None


def start(kind: str, label: str, fn: Callable[[JobCtx], str | None]) -> tuple[Job, bool]:
    """Run `fn(ctx)` on a daemon thread. Returns `(job, reused)`.

    `fn` returns the summary line, or None. It should call `ctx.check()` wherever stopping is safe
    and let `JobCancelled` propagate.

    ⚠⚠ STARTING IS IDEMPOTENT PER (kind, label), AND WITHOUT THAT CANCEL CANNOT WORK. Nothing used
    to stop a second press launching a second identical job: two fills over the same 1,712
    constituents, sharing one global rate limiter so both crawl, and a Cancel that stops exactly one
    of them. From the outside that is a Cancel button that does nothing — the run "keeps going",
    because a different run is still going.

    It is not a hypothetical press. `PortfolioFundamentalsRefresh` only knows it has a job in flight
    from its own React state, so reopening the modal — or reloading the page — brings the button
    back reading "Refresh benchmark" while the work is still running. Pressing it again is the
    obvious thing to do and was the wrong thing to do.

    ⚠ ATTACH, DO NOT REFUSE. Returning an error would be correct and useless: the reader wants the
    thing they asked for, and it is already happening. Handing back the running job means the second
    press adopts it, the button flips to Cancel, and the UI heals itself.

    ⚠ THE KEY IS (kind, label) BECAUSE THAT IS WHAT "THE SAME WORK" MEANS HERE — two different
    companies, baskets or indices differ in `label`; the same one twice is a duplicate. A caller that
    genuinely needs concurrent same-label runs would need a distinguishing label, which is the honest
    way to express it.
    """
    existing = find_running(kind, label)          # reaps stalled jobs on the way — see above
    _prune()
    if existing is not None:
        _log.info("[job] %s (%s) is already running as %s — attaching", label, kind, existing.id)
        return existing, True
    job = Job(id=uuid.uuid4().hex[:12], kind=kind, label=label)
    with _REGISTRY_LOCK:
        _JOBS[job.id] = job
    ctx = JobCtx(job)

    def _runner() -> None:
        try:
            summary = fn(ctx)
            job.summary = summary
            job.status = "done"
            ctx.emit("done", summary or "finished")
        except JobCancelled as e:
            job.status = "cancelled"
            # The worker's own account of where it stopped when it gave one — see `JobCancelled`.
            job.summary = str(e) or "cancelled — stopped at a safe point"
            # ⚠ NAMED AS AN OUTCOME, NOT AN ERROR. A cancelled job did what it was told; rendering
            # it in red beside a genuine failure teaches the reader to ignore both.
            ctx.emit("cancelled", job.summary)
        except Exception as e:  # noqa: BLE001
            _log.warning("[job] %s (%s) failed — %s: %s", label, kind, type(e).__name__, e)
            job.status = "failed"
            job.summary = f"{type(e).__name__}: {str(e)[:160]}"
            ctx.emit("error", job.summary)
        finally:
            job.ended_at = time.time()

    threading.Thread(target=_runner, name=f"job-{job.id}", daemon=True).start()
    return job, False


def get(job_id: str) -> Job | None:
    with _REGISTRY_LOCK:
        return _JOBS.get(job_id)


def cancel(job_id: str) -> Job | None:
    """Request cancellation. Idempotent; returns None for an unknown id.

    ⚠ IT ONLY *REQUESTS*. The job stops when its worker next reaches a `ctx.check()`, which is why
    the response reports `cancel_requested` rather than a status of `cancelled`."""
    job = get(job_id)
    if job is None:
        return None
    if not job.terminal:
        job._cancel.set()  # noqa: SLF001
    return job


def listing() -> list[dict]:
    """Everything still in the registry, newest first — running jobs plus recently finished ones,
    which is what lets a page reload re-attach to work already in flight."""
    _prune()
    with _REGISTRY_LOCK:
        jobs = sorted(_JOBS.values(), key=lambda j: j.created_at, reverse=True)
    return [j.public() for j in jobs]

File: frontend/lib/stores/jobs.ts
import { apiFetch } from '../apiFetch';
import { API_URL } from '../apiUrl';
import { traceError } from '../debugTrace';
import { createStore } from '../store';
import { runSSE } from '../stream';

/**
 * BACKGROUND JOBS, AND THE TOASTS THAT REPORT THEM.
 *
 * ⚠ MODULE-SCOPED ON PURPOSE — see `lib/store.ts`'s own note. The state lives outside React's
 * tree, so a job keeps reporting when the panel that started it unmounts and when the reader
 * navigates to another page. A toast owned by the component that launched it would vanish with
 * the route change while the work carried on invisibly, which is the failure this whole layer
 * exists to remove.
 *
 * The server is the authority on a job's state (`backend/jobs.py`); this holds only what is
 * needed to draw it, plus `dismissed`, which is a purely local "I have read this".
 */

export type JobStatus = 'running' | 'done' | 'failed' | 'cancelled';

export type JobToast = {
  id: string;
  /** What the reader pressed — a company name, not the ISIN the server keys on. */
  title: string;
  /**
   * The server's identity for this piece of work — `fundamentals.index` + `ACWI`.
   *
   * ⚠ IT IS HERE SO A CONTROL CAN FIND ITS OWN RUN AGAIN. A button knows it has a job in flight
   * only from its own React state, so reopening a modal or reloading the page brings it back
   * offering to START one while the work is still going. `(kind, label)` is what the server
   * de-duplicates on (`jobs.start`), so it is also the only key a component can match on to adopt
   * the run instead of launching a second.
   *
   * Empty until the first `job` frame arrives; `attachRunningJobs` fills it immediately.
   */
  kind: string;
  label: string;
  status: JobStatus;
  done: number;
  total: number;
  /** The latest progress line. One line: anything with detail goes to the console. */
  message: string;
  summary: string | null;
  /** Metered external calls spent (GuruFocus quota today). ⚠ 0 RENDERS NOTHING rather than "0
   *  calls" — a refusal and a cache hit both legitimately cost nothing, and a zero on every such
   *  card trains the eye to skip the number on the cards where it matters. */
  apiCalls: number;
  /** ⚠ NOT `status === 'cancelled'`. Cancellation is cooperative, so this is true from the moment
   *  the button is pressed while the job is still finishing its current feed. The card reads
   *  "cancelling…" in that window; without it the button would look inert for several seconds. */
  cancelRequested: boolean;
  dismissed: boolean;
};

type JobsState = { jobs: JobToast[] };

export const jobsStore = createStore<JobsState>({ jobs: [] });

/**
 * How long a finished card lingers before it fades, in seconds — the countdown the card shows.
 *
 * ⚠ THE THREE ARE NOT THE SAME NUMBER, because they are not equally worth reading. A success is
 * confirmation of something you asked for and already expected; a cancellation is worth a glance
 * to see how far it got; a failure is the only one carrying information you did not have, and it
 * points at the console for the rest. Giving them one duration would either rush the failure off
 * screen or leave successes piling up in the corner.
 *
 * ⚠ THE COUNTDOWN IS VISIBLE AND PAUSES ON HOVER (see `JobCard`). A toast that vanishes while
 * being read is worse than one that never appeared — you know you missed something and cannot get
 * it back.
 */
export const LINGER_SECONDS: Record<JobStatus, number> = {
  running: 0,        // not stale yet — no countdown
  done: 6,
  cancelled: 8,
  failed: 12,
};

const upsert = (id: string, patch: Partial<JobToast>) => {
  jobsStore.set((s) => {
    const i = s.jobs.findIndex((j) => j.id === id);
    if (i < 0) return {};
    const next = s.jobs.slice();
    next[i] = { ...next[i], ...patch };
    return { jobs: next };
  });
};

export const dismissJob = (id: string) =>
  jobsStore.set((s) => ({ jobs: s.jobs.filter((j) => j.id !== id) }));

/**
 * Ask the server to stop. ⚠ It halts at its next safe point, not immediately — the card shows
 * `cancelRequested` straight away so the press is acknowledged, and `status` follows when the
 * worker actually stops.
 *
 * ⚠ RETURNS WHETHER THE REQUEST LANDED, because a caller that puts its own button into a
 * "Cancelling…" state has to be able to take it back out again. Swallowing the failure here left
 * such a button disabled and lying for the rest of the run — the request never arrived, the job
 * carried on, and the only control that could stop it had turned itself off. Additive: every
 * existing call site ignores the value and is unchanged.
 */
export async function cancelJob(id: string): Promise<boolean> {
  upsert(id, { cancelRequested: true, message: 'cancelling…' });
  // ⚠⚠ LOCAL JOBS FIRST, AND THIS IS WHY THERE IS STILL ONE CANCEL PATH. `JobToaster`'s Cancel
  // button calls this function and nothing else; a second kind of job with a second kind of cancel
  // would mean teaching that button which sort it is looking at, at which point every future
  // control has to know too. A job that runs in this tab has no `/api/jobs/{id}` to POST to — the
  // POST would 404 and the card would sit at "cancelling…" for ever — so it is aborted here.
  const abort = localCancels.get(id);
  if (abort) { abort(); return true; }
  try {
    const r = await apiFetch(`${API_URL}/api/jobs/${encodeURIComponent(id)}/cancel`,
      { method: 'POST' });
    if (!r.ok) throw new Error(`HTTP ${r.status}`);
    return true;
  } catch (e) {
    traceError('jobs', `could not cancel job ${id}`, e);
    upsert(id, { cancelRequested: false, message: 'cancel failed — see the console' });
    return false;
  }
}

const isTerminal = (s: JobStatus) => s !== 'running';

/** Abort handles for jobs with no server side, keyed by toast id — see `startLocalJob`. */
const localCancels = new Map<string, () => void>();

/**
 * Report a piece of work THIS TAB is doing on the shared toast stack.
 *
 * ⚠⚠ NOT EVERY CANCELLABLE THING IS A SERVER JOB. `startJob` needs an endpoint that owns the work
 * and hands back a `job_id`; re-reading a cached GET has no such endpoint and does not deserve
 * one. Without this, a control like the Deep Valuation tab's share-price refresh had two bad
 * options: paint its own private spinner — a second progress vocabulary the reader has to learn,
 * and the exact thing the job layer was built to delete — or invent a backend job for a fetch.
 *
 * ⚠ CANCELLATION IS AN `AbortController`, AND IT IS REAL. `cancelJob` finds this handle before it
 * reaches for the network, so the toaster's own Cancel button works on these with no change to it.
 *
 * ⚠ IT DIES WITH THE TAB, AND THAT IS THE ONE THING A SERVER JOB DOES BETTER. A route change or a
 * reload takes the work with it — there is nothing to re-attach to (`attachRunningJobs` lists the
 * server's jobs, and this is not one). Use it only for work short enough that losing it costs
 * nothing; anything that outlives a page view belongs on the server.
 *
 * `run` returns the card's summary line. Throwing marks the card failed with the message.
 */
export function startLocalJob(
  title: string, kind: string, run: (signal: AbortSignal) => Promise<string | void>,
): string {
  const id = `local:${kind}:${
    typeof crypto !== 'undefined' && 'randomUUID' in crypto ? crypto.randomUUID() : `${performance.now()}`}`;
  const ctrl = new AbortController();
  localCancels.set(id, () => ctrl.abort());
  jobsStore.set((st) => ({
    jobs: [...st.jobs, {
      id, title, kind, label: title, status: 'running' as JobStatus, done: 0, total: 0,
      message: '', summary: null, apiCalls: 0, cancelRequested: false, dismissed: false,
    }],
  }));
  void (async () => {
    try {
      const summary = await run(ctrl.signal);
      // ⚠ THE SIGNAL, NOT THE ERROR, DECIDES. A cancelled fetch can resolve rather than throw
      // (a cacheable read is shared, so aborting one caller does not stop the request), and a card
      // that went green on a run the reader stopped is worse than one that never reported.
      upsert(id, ctrl.signal.aborted
        ? { status: 'cancelled', summary: 'cancelled', message: '' }
        : { status: 'done', summary: summary || null, message: '' });
    } catch (e) {
      upsert(id, ctrl.signal.aborted
        ? { status: 'cancelled', summary: 'cancelled', message: '' }
        : {
          status: 'failed',
          summary: e instanceof Error ? e.message : String(e),
          message: 'see the console',
        });
      if (!ctrl.signal.aborted) traceError('jobs', `local job ${kind} failed`, e);
    } finally {
      localCancels.delete(id);
    }
  })();
  return id;
}

/**
 * Follow one job's stream to its end.
 *
 * `after` lets a re-attach ask for the events it missed, so a page reload shows the run's history
 * rather than joining mid-sentence.
 */
/**
 * `onProgress` — called for every progress line, so a caller can refresh what it shows AS the job
 * runs rather than only when it ends.
 *
 * ⚠ IT IS NOT THE TOAST'S JOB. The toast already narrates; this exists for callers whose SCREEN is
 * the thing the job changes — "Refresh all" rewrites 44 account rows over several minutes, and
 * repainting them only at the end means staring at figures you have already replaced. Optional, so
 * every existing caller is untouched.
 *
 * ⚠ IT MUST NOT THROW. It runs inside the stream handler; an exception here would kill the
 * progress stream and the toast with it, turning a cosmetic nicety into a lost job. Wrapped below.
 */
export function watchJob(id: string, title: string, after = 0,
                         seed?: { kind?: string; label?: string },
                         onProgress?: (e: { done?: number; total?: number; message?: string })
                           => void): Promise<JobToast> {
  jobsStore.set((s) => (s.jobs.some((j) => j.id === id) ? {} : {
    jobs: [...s.jobs, {
      id, title, kind: seed?.kind ?? '', label: seed?.label ?? '',
      status: 'running' as JobStatus, done: 0, total: 0,
      message: 'starting…', summary: null, apiCalls: 0, cancelRequested: false, dismissed: false,
    }],
  }));

  return new Promise<JobToast>((resolve) => {
    // ⚠ NO AUTO-DISMISS TIMER HERE. The card owns the countdown, because only the card knows
    // whether the reader is hovering it — a timer started from this side would fire regardless and
    // yank the toast out from under the cursor. This just reports the outcome.
    const finish = () => {
      const job = jobsStore.get().jobs.find((j) => j.id === id);
      resolve(job ?? {
        id, title, kind: seed?.kind ?? '', label: seed?.label ?? '',
        status: 'failed', done: 0, total: 0, message: '', summary: null, apiCalls: 0,
        cancelRequested: false, dismissed: false,
      });
    };

    void runSSE(
      `${API_URL}/api/jobs/${encodeURIComponent(id)}/stream?after=${after}`,
      { method: 'GET' },
      (raw) => {
        const e = raw as {
          type?: string; kind?: string; label?: string; message?: string; status?: JobStatus;
          done?: number; total?: number; summary?: string | null; cancel_requested?: boolean;
          api_calls?: number;
        };
        if (e.type === 'job') {
          upsert(id, {
            status: e.status ?? 'running',
            done: e.done ?? 0,
            total: e.total ?? 0,
            summary: e.summary ?? null,
            apiCalls: e.api_calls ?? 0,
            cancelRequested: !!e.cancel_requested,
            // ⚠ ONLY FROM THE `job` FRAME. A progress EVENT also carries a `kind` and it means
            // something entirely different there (`progress` / `skip` / `error`) — writing that
            // into the identity would make every control lose track of its own run one line in.
            ...(e.kind ? { kind: e.kind } : {}),
            ...(e.label ? { label: e.label } : {}),
          });
          return;
        }
        // A progress line. `done`/`total` ride along on the ones that have them; the others are
        // pure narration and must not reset the bar to zero.
        upsert(id, {
          message: e.message ?? '',
          ...(typeof e.done === 'number' ? { done: e.done } : {}),
          ...(typeof e.total === 'number' ? { total: e.total } : {}),
        });
        // ⚠ AFTER the toast, and swallowed. The toast is the job's own record and must land even
        // if a listener misbehaves; see the note on `onProgress`.
        if (onProgress) {
          try {
            onProgress({ done: e.done, total: e.total, message: e.message });
          } catch (err) {
            traceError('jobs', `an onProgress listener for ${id} threw`, err);
          }
        }
      },
    ).then(finish).catch((err) => {
      traceError('jobs', `the stream for job ${id} failed`, err);
      upsert(id, { status: 'failed', message: 'lost the progress stream — see the console' });
      finish();
    });
  });
}

/**
 * Start a job and follow it.
 *
 * `startUrl` is the endpoint that OWNS the work and returns `{job_id}` — there is deliberately no
 * generic "run this kind" starter, on the server or here.
 *
 * Returns the id plus a promise that settles when the job reaches a terminal state, so a caller
 * can refresh its own view afterwards without knowing anything about toasts.
 */
export async function startJob(
  startUrl: string, title: string, init?: RequestInit,
  /** Per-progress-line callback — see `watchJob`. For callers whose screen the job rewrites. */
  onProgress?: (e: { done?: number; total?: number; message?: string }) => void,
): Promise<{ id: string; done: Promise<JobToast>; body: Record<string, unknown> }> {
  // ⚠ `init` IS OPTIONAL AND MERGED AFTER `method`, so a caller can add a JSON body (the basket
  // fill posts its holdings) without being able to turn this into a GET by accident.
  const r = await apiFetch(startUrl, { ...init, method: 'POST' });
  if (!r.ok) throw new Error(`HTTP ${r.status}`);
  const b = (await r.json()) as { job_id: string };
  // ⚠ THE START RESPONSE IS HANDED BACK, NOT JUST THE HANDLE. An endpoint often knows something
  // at start-up that the job itself never reports — the portfolio fill returns how many holdings
  // it could reach and why the rest it could not, which the caller has to render BEFORE the first
  // progress line arrives. Discarding it forced a second request for data we already had.
  // Additive: every existing caller destructures `{ id, done }` and is untouched.
  return { id: b.job_id, done: watchJob(b.job_id, title, 0, undefined, onProgress),
    body: b as Record<string, unknown> };
}

/**
 * ⚠ THE ONE CALL THAT MAKES RUNNING WORK VISIBLE MUST NOT BE A SINGLE SHOT.
 *
 * `attachRunningJobs` fires once, from a `useEffect` in the root layout, the moment the role
 * resolves — so a backend that is unreachable for that one second costs the reader every toast
 * until they reload the whole tab, while the server carries on working. And "unreachable for one
 * second" is the normal case, not the exotic one: a `uvicorn --reload` restart on a file save, a
 * Railway redeploy, a laptop waking up. All of them surface as `TypeError: Failed to fetch` — the
 * request left, the connection died, no status was ever returned.
 *
 * Bounded and backing off, because a backend that is genuinely down must not be polled forever:
 * ~13s of trying, then it gives up and says so in the console.
 */
const ATTACH_RETRY_MS = [1_000, 3_000, 9_000];

const _sleep = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

/** ⚠ THE RETRY MADE `attachRunningJobs` RE-ENTRANT. It used to be over in one round trip, so a
 *  second call could not overlap the first; now one can still be sleeping when the effect refires
 *  (the admin flag resolving, a "view as user" toggle), and two loops would each adopt the same
 *  jobs in the window before `watchJob` puts them in the store. */
let _attaching = false;

/**
 * Re-attach to whatever is already running on the server.
 *
 * ⚠ THIS IS THE PAYOFF OF PUTTING A JOB IN THE MIDDLE. Without it a reload leaves the work running
 * with nothing on screen to say so — the exact state the old thread-and-queue endpoints left you
 * in, except now it is recoverable.
 */
export async function attachRunningJobs() {
  if (_attaching) return;                     // one attach at a time — see `_attaching`
  _attaching = true;
  try {
    for (let attempt = 0; ; attempt++) {
      try {
        const r = await apiFetch(`${API_URL}/api/jobs`);
        if (!r.ok) {
          // ⚠ 403 FOR A NON-ADMIN IS AN ANSWER, NOT AN ERROR — and so is a 404. A 5xx is not: it
          // is the server still coming up, which is the one non-answer worth asking again for.
          if (r.status >= 500 && attempt < ATTACH_RETRY_MS.length) {
            await _sleep(ATTACH_RETRY_MS[attempt]);
            continue;
          }
          return;
        }
        const rows = (await r.json()) as {
          id: string; kind: string; label: string; status: JobStatus;
        }[];
        for (const j of rows) {
          if (j.status !== 'running') continue;   // finished ones are history, not a toast
          if (jobsStore.get().jobs.some((x) => x.id === j.id)) continue;
          void watchJob(j.id, j.label, 0, { kind: j.kind, label: j.label });
        }
        return;
      } catch (e) {
        if (attempt >= ATTACH_RETRY_MS.length) {
          traceError('jobs', 'could not list running jobs', e);
          return;
        }
        await _sleep(ATTACH_RETRY_MS[attempt]);
      }
    }
  } finally {
    _attaching = false;
  }
}

export { isTerminal };
